src/libraries/Microsoft.PowerFx.Core/Binding/BinderUtils.cs

```diff
@@ -1265,22 +1291,47 @@ public static BinderCheckTypeResult CheckBinaryOpCore(IErrorContainer errorConta
 
                 case BinaryOp.Or:
                 case BinaryOp.And:
-                    var resLeftAnd = CheckTypeCore(errorContainer, leftNode, features, leftType, DType.Boolean, /* coerced: */ DType.Number, DType.Decimal, DType.String, DType.OptionSetValue, DType.UntypedObject);
-                    var resRightAnd = CheckTypeCore(errorContainer, rightNode, features, rightType, DType.Boolean, /* coerced: */ DType.Number, DType.Decimal, DType.String, DType.OptionSetValue, DType.UntypedObject);
+                    var resLeftAnd = CheckTypeCore(errorContainer, leftNode, features, leftType, DType.Boolean, /* coerced: */ DType.Number, DType.Decimal, DType.String, DType.UntypedObject);
+                    var resRightAnd = CheckTypeCore(errorContainer, rightNode, features, rightType, DType.Boolean, /* coerced: */ DType.Number, DType.Decimal, DType.String, DType.UntypedObject);
                     return new BinderCheckTypeResult() { Node = node, NodeType = DType.Boolean, Coercions = resLeftAnd.Coercions.Concat(resRightAnd.Coercions).ToList() };
 
                 case BinaryOp.Concat:
                     BinderCheckTypeResult resLeftConcat;
                     BinderCheckTypeResult resRightConcat;
+
+                    if (features.StronglyTypedBuiltinEnums)
+                    {
+                        if (leftType == DType.OptionSetValue && rightType == DType.OptionSetValue)
+                        {
+                            if (rightType.OptionSetInfo.EntityName == leftType.OptionSetInfo.EntityName &&
+                                rightType.OptionSetInfo.CanConcatenateStronglyTyped)
+                            {
+                                return new BinderCheckTypeResult() { Node = node, NodeType = leftType };
+                            }
+                        }
+                        else if (leftType == DType.OptionSetValue &&
+                                 leftType.OptionSetInfo.CanCoerceFromBackingKind && leftType.OptionSetInfo.CanConcatenateStronglyTyped)
```

Comment:
Would that allow something like `JSONFormat.IndentFour & ErrorKind.Div0`? I don't know if we should enable that

Reply:
It does.  I started off with not allowing this, but it gets odd fast as all option sets can coerce to string.  Take for example string interpolation, that is translated into concatenate:

```
>> IR( $"a {JSONFormat.IndentFour} {ErrorKind.Div0}" )
Concatenate:s("a ":s, FieldAccess(ResolvedObject(Microsoft.PowerFx.Core.Types.Enums.EnumSymbol), IndentFour), " ":s, FieldAccess(ResolvedObject(Microsoft.PowerFx.Core.Types.Enums.EnumSymbol), Div0))
```

What I settled on is that we would only concatenate strongly typed within the same enum.  So, yes you can do the above, but you can only use it in a text context:

```
>> JSON( 1, JSONFormat.IndentFour & MatchOptions.IgnoreCase )
Error 31-32: Invalid argument type (Text). Expecting a Enum (JSONFormat) value instead.
Error 31-32: The value 'i' is not supported as an option in the JSON function.
Error 0-4: The function 'JSON' has some invalid arguments.
```